rust/optimized-customhashmap/main.rs

```diff
@@ -48,9 +48,8 @@ fn try_main() -> Result<(), Box<dyn Error>> {
                     hash = FNV_OFFSET;
                 }
             } else {
-                if b'A' <= b && b <= b'Z' {
-                    buf[i] += b'a' - b'A';
-                }
+                // 0x20 (6th bit) is the only different bit between lowercase and uppercase
+                buf[i] |= 0x20;
```

Comment:
Thanks for trying this! Interestingly, I am actually able to observe a measurable improvement. Before:

```
$ hyperfine './rust/optimized-customhashmap/target/release/countwords < kjvbible_x10.txt'
Benchmark #1: ./rust/optimized-customhashmap/target/release/countwords < kjvbible_x10.txt
  Time (mean ± σ):     218.0 ms ±   9.9 ms    [User: 209.8 ms, System: 7.5 ms]
  Range (min … max):   210.1 ms … 244.0 ms    13 runs
```

After:

```
$ hyperfine './rust/optimized-customhashmap/target/release/countwords < kjvbible_x10.txt'
Benchmark #1: ./rust/optimized-customhashmap/target/release/countwords < kjvbible_x10.txt
  Time (mean ± σ):     201.1 ms ±   5.1 ms    [User: 193.6 ms, System: 7.3 ms]
  Range (min … max):   191.7 ms … 211.9 ms    15 runs
```

Reply:
@BurntSushi  Also thank you for contributing so many great solutions with detailed descriptions. I learned a lot from it!

Reply:
Hmmm, I'm not sure about this, because it also modifies some ASCII 

* \[ to {
* \\ to \|
* \] to \}
* \^ to \~
* \_ to 0x7f

It *happens* to work for the test input, because the Bible has no `@` or `^` in it :-), but I think we should revert it so it has a chance of handling more general inputs (we're not stripping punctuation).

Reply:
You might try this instead? Taken from std impl of `to_ascii_lowercase`:

```rust
buf[i] |= (self.is_ascii_uppercase() as u8) << 5;
```

Reply:
Let's just revert this now, or update to the `(buf[i].is_ascii_uppercase() as u8) << 5` method if that's faster.

Reply:
@benhoyt Updated to @BurntSushi suggestion: 
```rust
buf[i] |= (buf[i].is_ascii_uppercase() as u8) << 5;
```